### PR TITLE
Extend finditerator and add numMatches

### PR DESCRIPTION
**Pull request by julie-sullivan**
* created a new iterator that extends the existing one and includes `numMatches`
* moved the COUNT logic to `MongoDBNativeQuery` from `MongoDBCollection`
* SKIP_COUNT is removed and now has no effect. 
* The COUNT logic is correct now, and independent of SKIP and LIMIT.

### CODE REVIEW

File: commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBCollection.java [hunk 1]
```diff
@@ -226,21 +224,11 @@ private DataResult endWrite(long numMatches, long numInserted, long numUpdated,
...
-            ExecutorService executor = Executors.newSingleThreadExecutor();
-            countFuture = executor.submit(() -> mongoDBNativeQuery.count(clientSession, query));
-        }
-
-        /**
-         * Getting the cursor and setting the batchSize from options. Default value set to 20.
-         */
-        FindIterable<Document> findIterable = mongoDBNativeQuery.find(clientSession, query, projection, options);
+        MongoDBIterator<Document> findIterable = mongoDBNativeQuery.find(clientSession, query, projection, options);
         MongoCursor<Document> cursor = findIterable.iterator();
```
Comment (pfurio) [inline commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBCollection.java]: I don't think it is necessary doing findIterable.iterator(); The idea of having the MongoDBIterator is also to be able to iterate over the elements for you. I would remove the line 228, and would use findIterable.hasNext(); findIterable.next(); and findIterable.close();

File: commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBIterator.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package org.opencb.commons.datastore.mongodb;
+
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoCursor;
+
+import java.util.Iterator;
+
+public class MongoDBIterator<E> implements Iterator<E> {
```
Comment (pfurio) [inline commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBIterator.java]: I would implement Closeable as well to be able to close the iterator

File: commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBIterator.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        return iterator.iterator().next();
+    }
+
+    public FindIterable<E> getIterable() {
+        return iterator;
+    }
+
+    public MongoCursor<E> iterator() {
+        return iterator.iterator();
+    }
```
Comment (pfurio) [inline commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBIterator.java]: This method should not be necessary. We should call to .next() and .hasNext(). We would also need to implement close() method to close the iterator.

File: commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBNativeQuery.java [hunk 4]
```diff
@@ -99,11 +100,17 @@ public long count(ClientSession clientSession, Bson query) {
...
-    public FindIterable<Document> find(Bson query, Bson projection, QueryOptions options) {
+    public MongoDBIterator<Document> find(Bson query, Bson projection, QueryOptions options) {
         return find(null, query, projection, options);
     }
 
-    public FindIterable<Document> find(ClientSession clientSession, Bson query, Bson projection, QueryOptions options) {
+    public MongoDBIterator<Document> find(ClientSession clientSession, Bson query, Bson projection, QueryOptions options) {
+
+        Future<Long> countFuture = null;
+        if (options.getBoolean(QueryOptions.COUNT)) {
```
Comment (pfurio) [inline commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBNativeQuery.java]: Check options != null

File: commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBNativeQuery.java [hunk 5]
```diff
@@ -170,7 +177,15 @@ public long count(ClientSession clientSession, Bson query) {
             findIterable.maxTime(options.getLong(QueryOptions.TIMEOUT), TimeUnit.MILLISECONDS);
         }
 
-        return findIterable;
+        long numMatches = -1;
+        if (options.getBoolean(QueryOptions.COUNT)) {
```
Comment (pfurio) [inline commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBNativeQuery.java]: Check options != null

File: commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBIterator.java (hunk 3 above)
Comment (julie-sullivan) [reply commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoDBIterator.java]: Yes I agree! However, [MongoPersistentCursor.java](https://github.com/opencb/java-common-libs/blob/develop/commons-datastore/commons-datastore-mongodb/src/main/java/org/opencb/commons/datastore/mongodb/MongoPersistentCursor.java#L114) returns a `FindIterable<Document>`. Should I change to `MongoDBIterator`? I had a compile error with sort and batchsize so I'd need to add setters for those, I think?

**Pull request merged**